### ISSUE TITLE
Flex Attention and Nested Tensor: very high VRAM usage

### ISSUE THREAD

**Issue by mrinaldi97** [pytorch/pytorch#155065]
### 🐛 Describe the bug

Hello,
i am trying to write from scratch the code to train new transformers architecture and because I needed document mask as well as alibi bias, I decided to try to develop using FlexAttention + Nested (jagged) tensors to adopt recent PyTorch's developments and obtain a clean code.

Unfortunately, I am having issues in implementing the MultiHeadAttention in this way. 
First, I confirm the issue https://github.com/pytorch/pytorch/issues/154556: I cannot compile FlexAttention using Jagged tensors, and this already is a big problem because in this way performance is worst than using padded tensors. But I hope the solution will be found because I saw the issue is being worked on.

Second, and the reason I am filing this issue, I noticed a dramatic increase in memory usage when using FlexAttention + Nested tensors.
For example, a model working with tensors torch.Size([16, 12, j14, 64]), where j4 is at maximum 2048, is using 2,7GB on my 3090 at batch size 2, while doing the same using flex attention raises the occupation (measured with a simple torch.cuda.mem_get_info(device)) at 8GB. This means that higher batch size that creates no problem with FlexAttention and normal tensors are impossible to run without very large VRAM.
Here some other stats:
batch_size 16 SDPA Nested: 22GB
batch_size 16 Flex Nested: OOM
batch_size 4 SDPA Nested: 3810MB
batch_size 4 Flex Nested: OOM
batch_size 4 Flex Normal 8GB
batch_size 3 Flex Nested: 15GB
batch_size 3 Flex Normal: 6B
batch_size 3 SDPA Nested: 3164GB
batch_size 3 SDPA Normal: 3166GB
batch_size 2 SDPA Nested: 2,7GB  
batch_size 2 Flex Nested: 8GB          
batch_size 2 SDPA Normal: 2,7GB 
batch_size 2 Flex Normal: 5,7GB

It seems that Flex+Nested scales very poorly with batch size. I am not very experienced, so I don't know if the issue lies in Flex Attention or in my attention implementation.
Here is a stripped down version of my attention function, I tried to cut as many extra stuff as possible to keep the code short. Alibi and block_mask code is omitted, because it has no influence on memory consumption in my experiments. The implementation is similar to the ones of Pytorch's tutorials.

What I noticed is that the OOM is caused in the function _math_attention_inner in torch/_higher_order_ops/flex_attention.py where it computes     scores = (query @ key.transpose(-2, -1)).to(dtype=working_precision).
I tried to get the shape of the tensor at that point, and I realized that at that point of the computation, the query tensor is no longer a nested tensor but a normal tensor. May be this the reason of the high memory usage? I tried to look for functions that converted the nested tensor to a padded/unbinded tensor in torch/_higher_order_ops/flex_attention.py and torch/nn/attention/flex_attention.py but I had no luck in finding them. Where does the conversion happen?

Is it related to my implementation or is an issue with how Flex Attention works with Nested Tensors?
Thank you!

```python
class MultiHeadAttention(nn.Module):
    def __init__(
                self,
                q_dim: int,
                k_dim: int,
                v_dim: int,
                tot_dim: int,
                nheads: int,
                bias: bool,
                #dropout: float=0.0, //Not supported by FlexAttention yet
                ):
        super().__init__()
        self.nheads=nheads
        assert tot_dim % self.nheads == 0, "Embedding dim is not divisible by nheads"
        self.tot_dim=tot_dim
        self.bias=bias
        self.qkv_samedim = q_dim==k_dim and q_dim==v_dim
        self.residual_dim=q_dim 
        
        if self.qkv_samedim:
            self.packed_proj=nn.Linear(self.residual_dim,3*tot_dim,bias=bias)
        else:
            self.q_proj=nn.Linear(q_dim,tot_dim,bias=bias)
            self.k_proj=nn.Linear(k_dim,tot_dim,bias=bias)
            self.v_proj=nn.Linear(v_dim,tot_dim,bias=bias)
            
        self.out_proj=nn.Linear(tot_dim,self.residual_dim,bias=bias)
        self.head_dim=tot_dim//self.nheads
        
    def forward(
                self,
                query_input,
                key_input,
                value_input,
                block_mask = None
            ):

        #TODO: INVESTIGATE WHY .detach().requires_grad_() is required with nested tensors! With normal padded tensors, it works also without
        qlen=query_input.shape[1]
        klen=key_input.shape[1]
        if self.qkv_samedim:
            if query_input is key_input and key_input is value_input:
                result=self.packed_proj(query_input)
                query,key,value=torch.chunk(result,3,dim=-1)
            else:
                q_weight,k_weight,v_weight = torch.chunk(
                        self.packed_proj.weight,3,dim=0)
                if self.bias:
                    q_bias,k_bias,v_bias = torch.chunk(
                            self.packed_proj.bias,3,dim=0)
                else:
                    q_bias,k_bias,v_bias=None,None,None
                query=F.linear(query_input,q_weight,q_bias)
                key=F.linear(key_input,k_weight,k_bias)
                value=F.linear(value_input,v_weight,v_bias)
        
        else:
            query=self.q_proj(query_input)
            key=self.k_proj(key_input)
            value=self.v_proj(value_input)
        # (Batch, seqlen, tot_dim) => (Batch, seqlen, nhead, head_dim) => (Batch, nhead, seqlen, head_dim)
        query=query.unflatten(-1,[self.nheads, self.head_dim]).transpose(1,2).detach().requires_grad_()
        key=key.unflatten(-1,[self.nheads, self.head_dim]).transpose(1,2).detach().requires_grad_()
        value=value.unflatten(-1,[self.nheads, self.head_dim]).transpose(1,2).detach().requires_grad_()
        attn_output = flex_attention(query, key, value) #I removed block mask and score mod for readability of the issue
        #attn_output=scaled_dot_product_attention(query, key, value, attn_mask=None, is_causal=False)
        attn_output=attn_output.transpose(1,2).flatten(-2)
        return self.out_proj(attn_output)
```

### Versions

PyTorch version: 2.7.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A
Nvidia driver version: 570.133.07

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.8.3.14
[pip3] nvidia-cuda-cupti-cu12==12.8.57
[pip3] nvidia-cuda-nvrtc-cu12==12.8.61
[pip3] nvidia-cuda-runtime-cu12==12.8.57
[pip3] nvidia-cudnn-cu12==9.7.1.26
[pip3] nvidia-cufft-cu12==11.3.3.41
[pip3] nvidia-curand-cu12==10.3.9.55
[pip3] nvidia-cusolver-cu12==11.7.2.55
[pip3] nvidia-cusparse-cu12==12.5.7.53
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.8.61
[pip3] nvidia-nvtx-cu12==12.8.55
[pip3] pytorch-lightning==2.5.1.post0
[pip3] torch==2.7.0+cu128
[pip3] torch-atlas-ds==0.1.0
[pip3] torchaudio==2.7.0+cu128
[pip3] torchmetrics==1.7.1
[pip3] torchvision==0.22.0+cu128
[pip3] triton==3.3.0
[conda] Could not collect

**Comment by drisspg** [comment]
cc @jbschlosser

**Comment by mrinaldi97** [comment]
Thanks for the interest!
I am sorry that I suspect last time I haven't filed a very nice and understandable issue, so I prepared a minimal repro to test only the attention part with random tensors. The code is largely inspired to this [Pytorch's official blog tutorial](https://docs.pytorch.org/tutorials/intermediate/transformer_building_blocks.html). 

Now, I am on a research scholarship and we are starting the process of training a middle sized model. We wanted to create a SOTA codebase by using flexattn + nested tensors but if this problem is not solved we will have to use other methods. Two colleagues that are more expert than me in pytorch programming are also checking this issue.

Here's the results:
![Image](https://github.com/user-attachments/assets/753f5a2f-0e39-47ad-ad76-97964f5dabb7)





N=2: Mem(GB) SDPA=0.0 Flex=0.0 Flash=0.0 SDPA_nested=0.0 Flex_nested=0.0
N=4: Mem(GB) SDPA=0.0 Flex=0.0 Flash=0.0 SDPA_nested=0.0 Flex_nested=0.0
N=8: Mem(GB) SDPA=0.0 Flex=0.0 Flash=0.0 SDPA_nested=0.0 Flex_nested=0.0
N=16: Mem(GB) SDPA=0.0 Flex=0.0 Flash=0.0 SDPA_nested=0.0 Flex_nested=0.0
N=32: Mem(GB) SDPA=0.0 Flex=0.1 Flash=0.0 SDPA_nested=0.0 Flex_nested=0.3
N=64: Mem(GB) SDPA=0.1 Flex=0.4 Flash=0.1 SDPA_nested=0.0 Flex_nested=1.4
N=128: Mem(GB) SDPA=0.1 Flex=1.5 Flash=0.1 SDPA_nested=0.1 Flex_nested=4.2
N=256: Mem(GB) SDPA=0.4 Flex=8.7 Flash=0.4 SDPA_nested=0.3 Flex_nested=20.2
N=272: Mem(GB) SDPA=0.2 Flex=2.7 Flash=0.2 SDPA_nested=0.2 Flex_nested=22.3
N=288: Mem(GB) SDPA=0.2 Flex=2.4 Flash=0.2 SDPA_nested=0.2 Flex_nested=21.9
N=336: Mem(GB) SDPA=0.3 Flex=2.4 Flash=0.3 SDPA_nested=0.2 Flex_nested=inf
N=384: Mem(GB) SDPA=0.4 Flex=4.9 Flash=0.4 SDPA_nested=0.3 Flex_nested=inf
N=512: Mem(GB) SDPA=0.5 Flex=8.0 Flash=0.5 SDPA_nested=0.4 Flex_nested=inf
N=1024: Mem(GB) SDPA=1.0 Flex=12.3 Flash=1.0 SDPA_nested=0.8 Flex_nested=inf
N=2048: Mem(GB) SDPA=2.5 Flex=inf Flash=2.5 SDPA_nested=2.0 Flex_nested=inf
N=4096: Mem(GB) SDPA=5.8 Flex=inf Flash=5.8 SDPA_nested=4.5 Flex_nested=inf
N=8192: Mem(GB) SDPA=11.3 Flex=inf Flash=11.3 SDPA_nested=8.8 Flex_nested=inf
N=16384: Mem(GB) SDPA=inf Flex=inf Flash=inf SDPA_nested=9.8 Flex_nested=inf
N=32768: Mem(GB) SDPA=inf Flex=inf Flash=inf SDPA_nested=12.0 Flex_nested=inf

And here's the code to compute memory usage and plot graph about time and memory:

```python
import torch
from torch.nn.functional import scaled_dot_product_attention
from torch.nn.attention.flex_attention import flex_attention
from flash_attn import flash_attn_func
import timeit
import matplotlib.pyplot as plt
import numpy as np

def benchmark(func, *args, **kwargs):
    torch.cuda.synchronize()
    torch.cuda.reset_peak_memory_stats()
    begin = timeit.default_timer()
    try:
        func(*args, **kwargs)
        torch.cuda.synchronize()
        return timeit.default_timer() - begin, torch.cuda.max_memory_allocated()
    except:
        return float('inf'), float('inf')

def zipf_sentence_lengths(batch_size):
    lengths = np.empty(batch_size, dtype=int)
    for i in range(batch_size):
        lengths[i] = 1
        while np.random.zipf(1.2) not in [3, 386, 858]:
            lengths[i] += 1
    return torch.tensor(lengths)

def gen_batch(N):
    lengths = zipf_sentence_lengths(N)
    E_h = 48
    tensors = []
    for _ in range(3):
        tensors.append(torch.nested.nested_tensor([torch.randn(l.item(), E_h, 16, dtype=torch.bfloat16, device="cuda") for l in lengths], layout=torch.jagged).transpose(1,2))
    return tensors

results = []
for N in [2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192]:
    torch.manual_seed(6)
    query, key, value = gen_batch(N)
    
    try:
        padded = [t.to_padded_tensor(0.0) for t in [query, key, value]]
        time_sdpa, mem_sdpa = benchmark(scaled_dot_product_attention, *padded)
        time_flex, mem_flex = benchmark(flex_attention, *padded)
        time_flash, mem_flash = benchmark(flash_attn_func, *[t.transpose(1,2) for t in padded])
    except:
        time_sdpa = time_flex = time_flash = float('inf')
        mem_sdpa = mem_flex = mem_flash = float('inf')
    
    time_sdpa_nested, mem_sdpa_nested = benchmark(scaled_dot_product_attention, query, key, value)
    time_flex_nested, mem_flex_nested = benchmark(flex_attention, query, key, value)
    
    results.append([N, time_sdpa, time_flex, time_flash, time_sdpa_nested, time_flex_nested, 
                   mem_sdpa/1e9, mem_flex/1e9, mem_flash/1e9, mem_sdpa_nested/1e9, mem_flex_nested/1e9])
    print(f"N={N}: Mem(GB) SDPA={mem_sdpa/1e9:.1f} Flex={mem_flex/1e9:.1f} Flash={mem_flash/1e9:.1f} SDPA_nested={mem_sdpa_nested/1e9:.1f} Flex_nested={mem_flex_nested/1e9:.1f}")

data = np.array(results)
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))
ax1.loglog(data[:,0], data[:,6], 'o-', label='SDPA')
ax1.loglog(data[:,0], data[:,7], 's-', label='Flex')
ax1.loglog(data[:,0], data[:,8], '^-', label='Flash')
ax1.loglog(data[:,0], data[:,9], 'v-', label='SDPA Nested')
ax1.loglog(data[:,0], data[:,10], 'd-', label='Flex Nested')
ax1.set_xlabel('Batch Size')
ax1.set_ylabel('Memory (GB)')
ax1.set_title('Memory Usage')
ax1.legend()
ax2.loglog(data[:,0], data[:,1], 'o-', label='SDPA')
ax2.loglog(data[:,0], data[:,2], 's-', label='Flex')
ax2.loglog(data[:,0], data[:,3], '^-', label='Flash')
ax2.loglog(data[:,0], data[:,4], 'v-', label='SDPA Nested')
ax2.loglog(data[:,0], data[:,5], 'd-', label='Flex Nested')
ax2.set_xlabel('Batch Size')
ax2.set_ylabel('Time (s)')
ax2.set_title('Execution Time')
ax2.legend()
plt.tight_layout()
plt.show()

**Comment by jbschlosser** [comment]
Hey @mrinaldi97, unfortunately, as you mentioned, there are known issues with torch.compile + flex attention + nested tensor and we are no longer prioritizing development of nested tensors, so it's unlikely we'll be able to address this for you soon. Apologies that the recent blog post misled you.

For handling ragged sequence lengths with flex attention + torch.compile, I suggest checking out the document masking example [here](https://github.com/pytorch-labs/attention-gym/blob/main/attn_gym/masks/document_mask.py). Please let us know if you still have unaddressed needs involving flex attention that are not covered there.
